src/prolong_restrict/pr_loops.hpp

```diff
@@ -72,6 +58,54 @@ GetLoopBoundsFromBndInfo(const Info_t &info, const int ckbs, const int cjbs, int
 // a device version, which requires the buffer cache device only,
 // and a version that automatically swaps between them depending on
 // the size of the buffer cache.
+
+template <int DIM, class Stencil, TopologicalElement FEL, TopologicalElement CEL>
+KOKKOS_INLINE_FUNCTION void InnerProlongationRestrictionLoop(
+    team_mbr_t &team_member, std::size_t buf, const BufferCache_t &info,
+    const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+    const IndexRange &kb, const IndexRange &jb, const IndexRange &ib) {
+  const auto &idxer = info(buf).prores_idxer[static_cast<int>(CEL)];
+  par_for_inner(inner_loop_pattern_ttr_tag, team_member, 0, 0, 0, 0, 0, idxer.size() - 1,
+                [&](const int, const int, const int ii) {
+                  const auto [t, u, v, k, j, i] = idxer(ii);
+                  Stencil::template Do<DIM, FEL, CEL>(
+                      t, u, v, k, j, i, ckb, cjb, cib, kb, jb, ib, info(buf).coords,
+                      info(buf).coarse_coords, &(info(buf).coarse), &(info(buf).fine));
+                });
+}
+
+template <int DIM, class Stencil, TopologicalElement... ELs, class... Args>
+KOKKOS_INLINE_FUNCTION void IterateInnerProlongationRestrictionLoop(Args &&...args) {
+  (
+      [&] {
+        if constexpr (Stencil::OperationRequired(ELs, TE::NN))
+          InnerProlongationRestrictionLoop<DIM, Stencil, ELs, TE::NN>(
+              std::forward<Args>(args)...);
```

Comment:
Feels like there should be a constexpr for all if or something... (This is a random thought. Do not go off and reimplement this)